ci/main.py

```diff
@@ -3,61 +3,130 @@
+handler.setLevel(logging.INFO)
+log.addHandler(handler)
+log.setLevel(logging.INFO)
 
 
-def build_upload_recipes(p):
+def build_upload_recipes(p, channel):
+    '''Build & upload recipes recursively in a directory.
+
+    Parameters
+    ----------
+    p : str
+        Path to the recipes.
+    channel : str
+        Anaconda channel where the packages will be uploaded.
+    '''
     for root, dirs, files in os.walk(p):
         has_recipe = 'meta.yaml' in files
         if not dirs and has_recipe:
             with open(os.path.join(root, 'meta.yaml')) as f:
-                pkg = yaml.load(f)['package']
-                name = pkg['name']
-                version = pkg['version']
+                meta = yaml.load(f)
+                name = meta['package']['name']
+                version = meta['package']['version']
+                try:
+                    build_number = meta['build']['number']
+                except KeyError:
+                    # Build number is 0 if not specified
+                    build_number = 0
+                if is_already_uploaded(name, version, build_number, channel):
+                    # Only new packages (either version or build_number)
+                    log.info("Skipping package: {0}-{1}".format(name, version))
+                    continue
                 build(name, version, root)
                 if os.environ['TRAVIS_SECURE_ENV_VARS'] == 'true':
-                    upload(name, version)
+                    upload(name, version, channel)
                 else:
-                    print("Uploading not available in Pull Requests")
+                    log.info("Uploading not available in Pull Requests")
 
 
 def build(name, version, root):
-    print("Building package: {0}-{1}".format(name, version))
+    '''Build a recipe.
+
+    Parameters
+    ----------
+    root : str
+        the directory path for the recipe.
+    '''
     # Quote is need in case the root path has spaces in it.
     build_cmd = 'conda build "%s"' % root
-    print('BUILDING COMMAND: {0}'.format(build_cmd))
+    log.info('Build: {0}'.format(build_cmd))
     check_call(build_cmd, shell=True)
 
 
-def upload(name, version):
+def is_already_uploaded(name, version, build_number, channel):
+    '''Check if we want to build & upload a package.
+
+    It checks package name, version and build number
+    sequentially to decide whether to build and upload it or not.
+
+    Parameters
+    ----------
+    name : str
+        Package name.
+    version : str
+        Package version.
+    build_number : int
+        Build number
+    channel : str
+        Anaconda channel to check the previous build.
+
+    Returns
+    -------
+    Bool
+        If a package in the channel has the same name, the same
+        version, and an equal or higher build number, then return
+        True; otherwise, return False.
+
+    '''
     check_cmd = ('conda search --json --override-channels '
                  '-c {0} --spec {1}={2}').format(
```

Comment:
I can add an explicit check, but if it is not found, it returns a json file that doesn't fulfil the requirements in the `all(...)` check, so it works well (i.e., no Python error is raised). Example reply:
```json
$ conda search --json --override-channels -c biocore --spec hmmer=3.1b3
{
  "error": "No packages found in current linux-64 channels matching: hmmer 3.1b3", 
  "error_type": "NoPackagesFound"
}
```